### PR TITLE
tag-expressions-go, cucumber-expressions-go: Thread safe stack

### PR DESCRIPTION
**Pull request by ankitpokhrel**
## Summary

Current implementation of the stack is not thread-safe and can lead to a data race if accessed concurrently. 

```bash
WARNING: DATA RACE
Read at 0x00c00000e8d0 by goroutine 39:
  github.com/cucumber/cucumber/tag-expressions/go.TestEvaluatableStackOperations.func2()
      /src/github.com/cucumber/cucumber/tag-expressions/go/stack.go:26 +0x1d4

Previous write at 0x00c00000e8d0 by goroutine 38:
  github.com/cucumber/cucumber/tag-expressions/go.TestEvaluatableStackOperations.func1()
      /go/src/github.com/cucumber/cucumber/tag-expressions/go/stack.go:48 +0x108
```

This PR 
- Adds a lock to the stack operations https://github.com/cucumber/cucumber/commit/fcfaf80ca2c06dae11ce66e622fcc596e252e99d
- Adds tests for the stack to make sure there is no data race https://github.com/cucumber/cucumber/commit/0665327083752257ff9bf2fcfed3d53cdc5a91fe
- And run tests with race detector on in pipeline https://github.com/cucumber/cucumber/commit/047b096485e2c8515f961ad8335b32c3aa85e7b4

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] The change has been ported to Java.
- [ ] The change has been ported to Ruby.
- [ ] The change has been ported to JavaScript.
- [ ] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mpkorstanje** [comment]
I didn't dig into the root cause of the race condition, but rather then locking I think it would be better to the shared state immutable. It's easier to understand and less error prone.

**Comment by mpkorstanje** [comment]
> Current implementation of the stack is not thread-safe and can lead to a data race if accessed concurrently.

Okay. Did a quick gander through the code and as far as I can tell implementations of `InterfaceStack` should never be accessed concurrently - so there is no need to add locking. 


Also for some reason we've made `InterfaceStack`, `StringStack`, ect all public. I'm not sure if that was intentional.

File: cucumber-expressions/go/default.mk [hunk 1]
```diff
@@ -88,7 +88,7 @@ dist_compressed/$(LIBNAME)-%: dist/$(LIBNAME)-%
 .tested: .deps $(GO_SOURCE_FILES) .go-tested
 
 .go-tested:
-	go test ./...
+	go test ./... -race
```
Comment (aslakhellesoy) [inline cucumber-expressions/go/default.mk]: All of the `**/go/default.mk` files in this monorepo are copies of `/.templates/go/default.mk`. This is the file you should modify. Then run `source scripts/functions.sh && rsync_files` from root to propagate the change to all the sub directories.

**Comment by ankitpokhrel** [comment]
Closing based on our discussion on slack

**Pull request closed without merging**